### PR TITLE
Retry when rate-limited

### PR DESCRIPTION
**Pull request by jesperes**
Algorithm is taken from [this AWS blog post](https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/).

### CODE REVIEW

File: src/bec.app.src [hunk 1]
```diff
@@ -15,5 +15,7 @@
   {env, [ {bitbucket_username, ""}
         , {bitbucket_password, ""}
         , {bitbucket_url, "http://localhost:8000" }
+        , {base_sleep_time, 100}
+        , {cap_sleep_time, 60000}
```
Comment (jesperes) [inline src/bec.app.src:19]: Ping @onno-vos-dev, WDYT about these defaults?

File: test/bitbucket_api_SUITE.erl [hunk 2]
```diff
@@ -69,6 +72,19 @@ init_per_testcase(get_wz_branch_reviewers_with_mandatory, Config) ->
...
+            {ok, {{"1.1", 429, "Rate limited"}, [], ""}}
+        end,
+  FunOk = fun(_, _, _, _) ->
+            {ok, {{"1.1", 200, "OK"}, [], ""}}
+        end,
+
+  NumRateLimits = 4,
+
+  ok = meck:expect(httpc, request, 4,
+                   meck:seq(lists:duplicate(NumRateLimits, FunRateLimited) ++ [FunOk])),
```
Comment (andreashasse) [inline test/bitbucket_api_SUITE.erl:87]: nice test!

File: src/bec.app.src (hunk 1 above)
Comment (onno-vos-dev) [reply src/bec.app.src:19]: I'm not sure if I'd retry indefinitely which is what this PR seems to be doing unless I'm overlooking this? :thinking: 

I'd personally set a max amount of retries to say 15 which with the configs you've currently set would mean retrying for about 5-6 minutes which I'd say should be more than sufficient. If you're still getting 429's after that, there's probably something worse going on.

In order to fiddle with the times one, you could play with:
```erlang
F = fun(N, BaseSleepTime, CapSleepTime) ->
      Temp = min(CapSleepTime, BaseSleepTime bsl N),
      Temp div 2 + rand:uniform(Temp div 2) 
    end.
lists:sum([ F(N, 100, 60_000) || N <- lists:seq(1,15)]).
```

**Review by andreashasse — APPROVED** [review]

**Review by onno-vos-dev — APPROVED** [review]
LGTM even though I'd personally add a max_retries configuration in order to not get stuck in some CI pipeline forever if there's 429's consistently. But if you feel that there's no need and you'd prefer to retry forever then I'm fine with that too :+1:

File: src/bec.app.src (hunk 1 above)
Comment (onno-vos-dev) [reply src/bec.app.src:19]: I'm not sure if I'd retry indefinitely which is what this PR seems to be doing unless I'm overlooking this? :thinking: 

I'd personally set a max amount of retries to say 15 which with the configs you've currently set would mean retrying for about 5-6 minutes which I'd say should be more than sufficient. If you're still getting 429's after that, there's probably something worse going on.

In order to fiddle with the times one, you could play with:
```erlang
F = fun(N, BaseSleepTime, CapSleepTime) ->
      Temp = min(CapSleepTime, BaseSleepTime bsl N),
      Temp div 2 + rand:uniform(Temp div 2) 
    end.
lists:sum([ F(N, 100, 60_000) || N <- lists:seq(1,15)]).
```

**Comment by jesperes** [comment]
> LGTM even though I'd personally add a max_retries configuration in order to not get stuck in some CI pipeline forever if there's 429's consistently. But if you feel that there's no need and you'd prefer to retry forever then I'm fine with that too +1

I'll add a `max_retries` config.

**Review by andreashasse — APPROVED** [review]

**Review by Taddic — APPROVED** [review]

**Review by onno-vos-dev — APPROVED** [review]

**Review by mikpe — APPROVED** [review]

**Review by onno-vos-dev — APPROVED** [review]

**Pull request merged**